pkg/controller/utils/duck_types.go

```diff
@@ -0,0 +1,33 @@
+package utils
+
+import (
+	"context"
+	"encoding/json"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// GetDuckType uses the client to fetch a type defined by gck and key, and
+// marshals it into the duck type defined in obj.
+func GetDuckType(ctx context.Context, client client.Client,
+	gvk schema.GroupVersionKind, key types.NamespacedName,
+	obj client.Object) error {
+	uObj := &unstructured.Unstructured{}
+	uObj.SetGroupVersionKind(gvk)
+	if err := client.Get(ctx, key, uObj); err != nil {
+		return err
+	}
+
+	// since client can be backed by cache,
+	// there is no way to really use the client-go .Into
+	// function to marshal only once form api response
+	// directly into the ducked type.
+	raw, err := uObj.MarshalJSON()
+	if err != nil {
+		return err
+	}
```

Comment:
Any extra validation we should do here to make sure it's correct and the user didn't make a mistake? I thought about checking for our required conditions but it's possible we just didn't add them yet. Should we validate certain fields are set that we expect?

Reply:
first, i don't think there is a need for us to validate. the clsuterinstall contract Go type will have the data when provided correctly by the implementation. if there is missing data, we will fail with can't find the data.
for example, imageSetRef, when the implementation didn't have it, we will see empty reference and we can error saying you didn't specify the imageSetRef ..

secondly, if we were interested in ensuring that the users set only implementations that are explicitly specified to fulfill the cluster install contract we can do that too using,


```
1. Add a label / annotation to the CRD of the implementation that specifies that it support clusterinstall contract. like `contracts.hive.openshift.io/<contract name (clusterinstall)> = true`
2. We can manage a list of contract => support GVKs locally and only accept the GVKs that support clusterinstall strategy
```
NOTE: we would need some one to have permissions to list CRDs to calculate this list. maybe the hive-operator ??